docs/proposals/Automated-Certificate-and-Key-Updates.md

```diff
@@ -0,0 +1,190 @@
+
+## Proposed solutions
+
+There are two methods to update keys or certificate automatically:
+- **Periodic retrieval of enabled keys or certificates**: Ratify periodically retrieve enabled keys or certificates from the KMS.
+  - Pros:
+    - Simplicity (no need for real-time event handling).
+    - Predictable resource usage.
+    - Works well for less time-sensitive use cases
+  - Cons:
+    - May lead to delays of updates.
+    - Frequent queries can impact performance.
+- **Event-Driven Notification**: Ratify subscribes to KMS events, when a relevant event occurs, Ratify receives an event notification. Then Ratify can retrieve the latest enabled keys or certificates from the KMS.
+    - Pros:
+      - Real-time responsiveness.
+      - Efficient use of resources (only fetches when needed).
+      - Minimizes downtime for key or certificates changes.
+    - Cons:
+      - Requires adapting to different event infrastructure (webhooks, message queues).
+      - Complexity in handling event delivery and retries.
+
+The proposed solution is **Periodic retrieval of enabled keys or certificates** because it is simpler and the update is not a time-sensitive action. Users can manual update keys or certificates as a complementary if required.
+
+## User experiences
+
+This section describes the experience that users interact with Ratify using the proposed solution. In summary, the proposed solution maintains the existing user experience for configuring `KMP` resources. Automatic updates of keys or certificates occur seamlessly in the background. However, if users utilize inline `KMP` resources, they will still need to manually update the keys or certificates. Importantly, the automatic updates do not prevent users from making manual updates when necessary.
+
+### Automatically update keys and certificates
+
+If users specify key or certificate versions in the `KMP` resource, then only specific versions are updated automatically. For example, if specific versions are disabled or deleted in KMS, then they cannot be used for signature verification after updates. See an example of `KMP` configuration with the version `${KEY_VER}` specified for a key `${KEY_NAME}`:
+
+```yaml
+apiVersion: config.ratify.deislabs.io/v1beta1
+kind: KeyManagementProvider
+metadata:
+name: keymanagementprovider-akv
+spec:
+type: azurekeyvault
+parameters:
+    vaultURI: https://${AKV_NAME}.vault.azure.net/
+    keys:
+    - name: ${KEY_NAME}
+        version: ${KEY_VER1}
+    tenantID: ${TENANT_ID}
+    clientID: ${IDENTITY_CLIENT_ID}
+```
+
+If users configure key or certificate names or aliases or IDs, no version are specified, then Ratify fetches all available versions and automatically update them. See an example of `KMP` configuration, no versions specified for the key `${KEY_NAME}`. Ratify will fetches all enabled key versions `${KEY_VER1}` and `${KEY_VER2}`, and update them automatically.
```

Comment:
@yizha1  As discussed in CC, I think we should consider enforcing last `n` versions instead of all. Azure Secret store CSI driver has a config `objectVersionHistory` which does something similar: https://azure.github.io/secrets-store-csi-driver-provider-azure/docs/getting-started/usage/#create-your-own-secretproviderclass-object